### ISSUE TITLE
Unserializing foreign key / m2m relations

### ISSUE THREAD

**Issue by tarpack** [etianen/django-reversion#508]
I'm using django-reversion to store revisions of a bunch of different models, all of which have some sort of relations (whether foreign keys, many to many, or some kind of reverse relationship from a different model). On some of these models, I care about storing the related objects in the revision so I'm using `follow` when I register the model, on others I don't. It's all working great.

I don't use the django admin, so through the low level API I eventually end up with a `reversion.models.Version` object. I want to show the user what this version looked like. Using `field_dict` gets me most of the way there, except for the relations. If the model has a `colors` property that is a many to many relationship to a `Color` model, the dictionary entry is going to be something like:

    {...,
     'colors': [1, 23, 54],
     ...}

That isn't useful to the user.

So my question is, is there some generic way that I could recognize that the integers values in the dictionary are supposed to be the primary keys of an object, and what models those objects are, so that when I'm building the template I can get the actual objects and show them to the user?

In this case I know that the `colors` dictionary key is going to be a list of primary keys to `Color` objects, so when I'm looping through the dictionary I can look for that key and take special action when I get to it. But I can't think of some generic way to do this, that doesn't rely on some previous knowledge of the model.

Basically I want some kind of "de-serializer" function that I can pass a `reversion.models.Version` object to and get back something sort of object-like (whether it is actually an object or just another dictionary) that has followed the relationships to fetch the actual objects.

This seems like a common problem, but I'm not sure what the solution is!

**Comment by etianen** [comment]
So, by introspecting Version.content_type.model_class()._meta.fields, you
can see what fields correspond to foreign keys and m2m relations.

You can get the version data for that referenced model by looking in
Version.revision.version_set.all(), which is a queryset of all other models
in the revision. With knowledge of the primary key value, and the
referenced model class from the introspected field, you can find the
Version of the referenced model in the queryset.

**Comment by tarpack** [comment]
Thanks. That is definitely helpful. I don't see m2m fields in Version.content_type.model_class()._meta.fields, but that gave me enough to go on to figure out I could get those at Version.content_type.model_class()._meta.many_to_many!

Is there documentation for all the properties of `reversion.models.Version` and `reversion.models.Revision`? When I first started registering models with `follow`, I couldn't figure out how to get to the other stored models. I reread the docs a few times, thinking I might have missed it, but ended up having to look through the code to learn about `reversion.models.Revision.version_set`.

**Comment by tarpack** [comment]
In case anybody else is looking at this, the django documentation on field access API is helpful: https://docs.djangoproject.com/en/1.9/ref/models/meta/#field-access-api

Instead of Version.content_type.model_class()._meta.fields and Version.content_type.model_class()._meta.many_to_many you can use Version.content_type.model_class()._meta.get_fields() to get all fields (including fk and m2m). And Version.content_type.model_class()._meta.get_field() can be used to get a field by name.

So if `version` is a `reversion.models.Version`, I can get what I want by doing something like:
```
for k,v in version.field_dict.iteritems():
    field = version.content_type.model_class()._meta.get_field(k)
    if type(field) is models.ManyToManyField:
        # loop through k, using each list entry as a primary key to objects using the field.related_model model
    elif type(field) is models.ForeignKey:
        # more or less the same thing as with m2m, but only one value!
    else:
        # the field isn't a relation, so probably just give the value to the user in its current state.
```

**Comment by etianen** [comment]
Nice!

Django-reversion could do with more docs about its mode structure. If I had
more time, I'd definitely do it.

If you feel like expanding the docs with information about the models,
please do go ahead with a pull request.
